core/src/main/java/com/datastax/oss/driver/api/core/config/DriverConfigLoader.java

```diff
@@ -133,6 +134,51 @@ static DriverConfigLoader fromUrl(URL url) {
         });
   }
 
+  /**
+   * Starts a builder that allows configuration options to be overridden programmatically.
+   *
+   * <p>Sample usage:
+   *
+   * <pre>{@code
+   * DriverConfigLoader loader =
+   *     DriverConfigLoader.programmaticBuilder()
+   *         .withDuration(DefaultDriverOption.REQUEST_TIMEOUT, Duration.ofSeconds(5))
+   *         .startProfile("slow")
+   *         .withDuration(DefaultDriverOption.REQUEST_TIMEOUT, Duration.ofSeconds(30))
+   *         .endProfile()
+   *         .build();
+   * }</pre>
+   *
+   * The resulting loader still uses the driver's default implementation (based on Typesafe config),
+   * except that the programmatic configuration takes precedence. More precisely, configuration
+   * properties are loaded and merged from the following (first-listed are higher priority):
+   *
+   * <ul>
+   *   <li>system properties
+   *   <li>properties that were provided programmatically
+   *   <li>{@code application.conf} (all resources on classpath with this name)
+   *   <li>{@code application.json} (all resources on classpath with this name)
+   *   <li>{@code application.properties} (all resources on classpath with this name)
+   *   <li>{@code reference.conf} (all resources on classpath with this name). In particular, this
+   *       will load the {@code reference.conf} included in the core driver JAR, that defines
+   *       default options for all mandatory options.
+   * </ul>
+   *
+   * Note that {@code application.*} is entirely optional, you may choose to only rely on the
+   * driver's built-in {@code reference.conf} and programmatic overrides.
+   *
+   * <p>The resulting configuration is expected to contain a {@code datastax-java-driver} section.
+   *
+   * <p>The loader will honor the reload interval defined by the option {@code
+   * basic.config-reload-interval}.
+   *
+   * <p>Note that the returned builder is <b>not thread-safe</b>.
+   */
+  @NonNull
+  static ProgrammaticConfigBuilder programmaticBuilder() {
```

Comment:
1. There's only so much we can do without leaking Typesafe config in the public API. I've tried to pick what seemed to be the most generic cases, but yes if you want programmatic _and_ URL-based it's not possible to compose, you'll have to switch to the internal API (and indeed use that constructor).
2. When we did the internal version there was a debate and we compromised on a crappy name. I want "programmatic" in the name (type and method) because that's what it is: the thing where you set options from your code instead of loading them from a file. Formally it should be `ProgrammaticDriverConfigLoaderBuilder` indeed, but that's a mouthful. So I shortened it to keep the bare minimum: it's programmatic, it relates to the config, and it's a builder. The fact that it produces a loader is implied by the return type of `build()`. I'm not hellbent on that, if you think it's not clear we can go back to the longer name.

Reply:
1. OK
2. Agreed to have "programmatic" in the name. What bothers me is that the current name reads to me as "a programmatic builder for configs" while it actually is a "programmatic builder for loaders of configs". I personally prefer the long name.

Reply:
OK I'll change the name.